things/service/src/main/java/org/eclipse/ditto/things/service/persistence/actors/strategies/commands/MigrateThingDefinitionStrategy.java

```diff
@@ -0,0 +1,365 @@
+            @Nullable final Metadata metadata) {
+
+        final Thing existingThing = getEntityOrThrow(thing);
+        final Instant eventTs = getEventTimestamp();
+        return handleMigrateDefinition(context, existingThing, eventTs, nextRevision, command, metadata);
+    }
+
+    private Result<ThingEvent<?>> handleMigrateDefinition(
+            final Context<ThingId> context,
+            final Thing existingThing,
+            final Instant eventTs,
+            final long nextRevision,
+            final MigrateThingDefinition command,
+            @Nullable final Metadata metadata) {
+
+        final DittoHeaders dittoHeaders = command.getDittoHeaders();
+        final boolean isDryRun = dittoHeaders.isDryRun();
+
+        // 1. Evaluate Patch Conditions and modify the migrationPayload
+        final JsonObject adjustedMigrationPayload = evaluatePatchConditions(
+                existingThing,
+                command.getMigrationPayload(),
+                command.getPatchConditions(),
+                dittoHeaders);
+
+        // 2. Generate Skeleton using definition and apply migration
+        final CompletionStage<Thing> updatedThingStage = generateSkeleton(command, dittoHeaders)
+                .thenApply(skeleton -> resolveSkeletonConflicts(
+                        existingThing, skeleton,
+                        command.isInitializeMissingPropertiesFromDefaults()))
+                .thenApply(mergedThing -> applyMigrationPayload(context,
+                        mergedThing, adjustedMigrationPayload, dittoHeaders, nextRevision, eventTs));
+
+        // 3. Validate and build event response
+        final CompletionStage<Pair<Thing, MigrateThingDefinition>> validatedStage = updatedThingStage
+                .thenCompose(mergedThing -> buildValidatedStage(command, existingThing, mergedThing)
+                        .thenApply(migrateThingDefinition -> new Pair<>(mergedThing, migrateThingDefinition)));
+
+        // If Dry Run, return a simulated response without applying changes
+        if (isDryRun) {
+            return ResultFactory.newQueryResult(
+                    command,
+                    validatedStage.thenApply(pair ->
+                            MigrateThingDefinitionResponse.dryRun(
+                                    existingThing.getEntityId().get(),
+                                    pair.first().toJson(),
+                                    dittoHeaders))
+            );
+        }
+
+        // 4. Apply migration and generate event
+        final CompletionStage<ThingEvent<?>> eventStage = validatedStage.thenApply(pair -> ThingMigrated.of(
+                pair.first(), nextRevision, eventTs, dittoHeaders,
+                metadata));
+
+        final CompletionStage<WithDittoHeaders> responseStage = validatedStage.thenApply(pair ->
+                appendETagHeaderIfProvided(command, MigrateThingDefinitionResponse.applied(existingThing.getEntityId().get(),
```

Comment:
`existingThing.getEntityId().get()` could be replaced with `context.getState()` - as `existingThing.getEntityId()` is optional, the `.get()` might throw an exception